### PR TITLE
Implement occ command to reset bruteforce attemps from a given IP address

### LINKED ISSUE #3058: Delete IP from oc_bruteforce_attempts via occ command
At the moment you have to delete an IP address which was blocked by a brute-force-detection via an SQL command on the database.

https://help.nextcloud.com/t/how-can-i-unblock-an-ip-blocked-through-brute-force-detection/5731

It would be nice to to this via an "occ" command.

### PR DESCRIPTION
**Pull request by joeried**
Adds a way to remove all bruteforce attempts (after the cutoff time) for a given IP address
Should resolve #3058 
I'm not entirely sure that the naming is correct.

@kesselb Please take a look.

### CODE REVIEW

**Comment by kesselb** [comment]
Thanks :+1: 


1. The autoloaders are not up to date. Please run: bash build/autoloaderchecker.sh and commit the result.
2. `php occ security:bruteforceattemps:reset-for-ip --ipadress=192.1.1.10` would be the command on my instance. That's quite long ;) I guess it's safe to use `security:bruteforce:unblock --ip` or something else. 
3. I don't get the part with the cut off. As administrator using that command I would expect that the ip is removed. Why the check for the cut off?

**Comment by joeried** [comment]
>     1. The autoloaders are not up to date. Please run: bash build/autoloaderchecker.sh and commit the result.
Thanks, I'll do that.

>     2. `php occ security:bruteforceattemps:reset-for-ip --ipadress=192.1.1.10` would be the command on my instance. That's quite long ;) I guess it's safe to use `security:bruteforce:reset --ip` or something else.
ipaddress is a argument and not an option. So your command would be `[...]security:bruteforceattemps:reset-for-ip 192.1.1.10` (if I'm not mistaken?) but of course thats still quite long and could be shortened.

>     3. I don't get the part with the cut off. As administrator using that command I would expect that the ip is removed. Why the check for the cut off?
The cut off is taken into account, as this would be consistent with the IP unblock after a successful login.
I thought the idea behind this was to not delete more entries than necessary to bypass the throttling.

File: lib/private/Security/Bruteforce/Throttler.php [hunk 1]
```diff
@@ -273,6 +273,24 @@ public function resetDelay($ip, $action, $metadata) {
...
 	}
 
+	/**
+	 * Reset the throttling delay for an IP address
+	 *
+	 * @param string $ip
+	 */
+	public function resetDelayForIP($ip){
+		$cutoffTime = (new \DateTime())
+			->sub($this->getCutoff(43200))
```
Comment (kesselb) [inline lib/private/Security/Bruteforce/Throttler.php]: Please make `43200` (and the other occurrence at resetDelay) a class constant. Alternatively add a method to calculate the cutoff time and use it by both methods.

File: lib/private/Security/Bruteforce/Throttler.php [hunk 2]
```diff
@@ -88,6 +88,17 @@ private function getCutoff($expire) {
 		return $d2->diff($d1);
 	}
 
+	/**
+	 *  Calculate the cut off timestamp
+	 *
+	 * @return int
+	 */
+	private function getCutoffTimestamp(){
```
Comment (kesselb) [inline lib/private/Security/Bruteforce/Throttler.php]:
```suggestion
	private function getCutoffTimestamp(): int {
```

**Comment by kesselb** [comment]
All right. Mind to do another rebase (to get rid of the composer annotation update - that's already landed in master) and squash the commits together?

**Review by kesselb — APPROVED** [review]

**Review by MorrisJobke — APPROVED** [review]
Code looks good 👍 Tested and works

**Comment by kesselb** [comment]
My bad. I overlooked the rebase and forgot to request a review earlier. Otherwise it might be landed for 19 already :see_no_evil:

**Comment by MorrisJobke** [comment]
I just pushed a fixed space - so let's merge this and continue on this :) Thanks for the feature :)

**Pull request merged**